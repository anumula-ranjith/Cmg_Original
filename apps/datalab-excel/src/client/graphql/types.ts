export type IQueryPath = Readonly<{
  [key: string]: string | IQueryPath;
}>;

export type GQLScalar = string | number | boolean | null;

export type GQLEnum = Record<string, GQLScalar>;

export type GQLNode = Readonly<{
  [key: string]: GQLScalar | GQLScalar[] | GQLNode | GQLNode[] | GQLEnum;
}>;

export type GQLPageInfo = Readonly<{
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}>;

export type GQLQueryType = Readonly<{
  pageInfo?: GQLPageInfo;
  totalCount?: number;
  items: GQLNode[];
}>;

export type GQLResponse = Readonly<{
  data?: Record<string, GQLQueryType | GQLSchema | GQLSchemaForTypeQuery> | GQLSchemaKind;
  errors?: GQLError[];
}>;

export type GQLLocation = Readonly<{
  line: number;
  column: number;
}>;

export type GQLErrorExtensions = Readonly<{
  type: string;
  field: string;
  responseName: string;
  specifiedBy: string;
}>;

export type GQLError = Readonly<{
  message: string;
  locations: GQLLocation[];
  extensions: GQLErrorExtensions;
}>;

export type GQLTypeRef = Readonly<{
  kind: string;
  name?: string;
  ofType?: GQLTypeRef;
  fields?: GQLField[];
}>;

export type GQLField = Readonly<{
  name: string;
  type: GQLTypeRef;
  description: string;
}>;

export type GQLType = Readonly<{
  kind: string;
  name: string;
  description?: string;
  fields?: GQLField[];
  interfaces?: GQLTypeRef[];
  possibleTypes?: GQLTypeRef[];
  enumValues?: { name: string; description?: string }[];
  inputFields?: GQLField[];
}>;

export type GQLSchema = Readonly<{
  queryType: GQLTypeRef;
  mutationType?: GQLTypeRef;
  subscriptionType?: GQLTypeRef;
  types: GQLType[];
  directives: {
    name: string;
    description?: string;
    locations: string[];
    args: GQLField[];
  }[];
}>;

export type GQLSchemaForTypeQuery = Readonly<{
  __type: {
    name: string;
    fields?: GQLField[];
  };
}>;

export type GQLSchemaKind = Readonly<{
  __schema: {
    name: string;
    kind: string;
    queryType: { fields: GQLField[] };
  };
}>;
