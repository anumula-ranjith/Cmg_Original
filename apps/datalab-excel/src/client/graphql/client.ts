import type { GraphQLSchema } from 'graphql';

import type { BuildQueryOptions } from './instrospection/build';
import { buildQuery } from './instrospection/build';
import type { GQLError, GQLQueryType, GQLResponse } from './types';

export async function fetchInstrospection(
  graphQlUrl: string,
  apiKey: string | null,
  query: string,
  variables: Record<string, string> | null = null
): Promise<GQLResponse> {
  if (!apiKey) {
    throw new Error('No API Key set');
  }
  const response = await fetch(graphQlUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      query: query,
      variables: variables,
    }),
  });
  if (response.status !== 200) {
    throw new Error(`Server responded with ${response.status} ${response.statusText}`);
  }

  const contentType = response.headers.get('content-type');
  if (!contentType || !contentType.includes('application/graphql-response+json')) {
    throw new Error(
      `Expected JSON response but received ${contentType ? contentType : 'no content type'}`
    );
  }
  const result = (await response.json()) as GQLResponse;
  if (result.errors) {
    const messages = result.errors.map((error: GQLError) => error.message).join('\n');
    throw new Error(`Errors: ${messages}`);
  }
  return result;
}

export async function fetchGraphqlData(
  schema: GraphQLSchema,
  graphQlUrl: string,
  apiKey: string | null,
  source: string,
  queryOptions: BuildQueryOptions,
  whereClause: string | null | undefined,
): Promise<GQLQueryType> {
  if (!apiKey) {
    throw new Error('No API key found');
  }
  const query = buildQuery(schema, source, queryOptions, whereClause);
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      query,
    }),
  };
  const response = await fetch(graphQlUrl, options);

  if (response.status !== 200) {
    throw new Error(`Server responded with ${response.status} ${response.statusText}`);
  }
  const result = (await response.json()) as GQLResponse;
  if (result.errors) {
    const messages = result.errors.map(error => error.message).join('<br />');
    throw new Error(`GraphQL: ${messages}`);
  }
  const typeData = (result.data ? result.data[source] : { items: [] }) as GQLQueryType;
  return typeData;
}
