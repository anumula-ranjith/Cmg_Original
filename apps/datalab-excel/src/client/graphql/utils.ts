import { findOperationRootTypeOrThrow } from '@cmg/graphql-toolkit-introspection';
import type { GraphQLOutputType, GraphQLSchema } from 'graphql';
import { GraphQLEnumType, getNamedType, isObjectType, isScalarType } from 'graphql';

import type { TableCell } from '../../excel/types';
import type { CoerceFieldsIntoOperationOptions } from './instrospection/types';
import type { GQLNode } from './types';

export const DEFAULT_MAX_NESTED_COERCION = 2;

export function escapeIfNeeded(s: string | object): string {
  let str: string;
  if (typeof s === 'object') {
    str = JSON.stringify(s).replace(/[{}"]/g, m => {
      switch (m) {
        case '{':
          return '\\{';
        case '}':
          return '\\}';
        case '"':
          return '\\"';
        default:
          return m;
      }
    });
  } else {
    str = s.toString();
  }
  return `"${str}"`;
}

export function stringifyResponse(response: unknown): TableCell {
  if (response === null || response === undefined) {
    return '';
  }
  if (typeof response === 'boolean') {
    return response ? 'TRUE' : 'FALSE';
  }
  if (typeof response === 'number') {
    return response;
  }
  if (typeof response === 'string') {
    return response.length < 1 ? '' : response;
  }
  if (typeof response === 'object') {
    if (Array.isArray(response)) {
      if (response.length === 0) {
        return '';
      }
    } else {
      return escapeIfNeeded(JSON.stringify(response));
    }
    return JSON.stringify(response);
  }
  return response as string;
}

export function parseGraphQlResponse(
  response: GQLNode[],
  userColumns: Record<string, string>,
  headers: string[]
): Record<string, TableCell>[] {
  const userColumnsKeys = Object.keys(userColumns);
  const result = response.map((item: GQLNode) => {
    const rows: Record<string, TableCell> = {};
    headers.forEach((header: string) => {
      if (userColumnsKeys.includes(header)) {
        rows[header] = userColumns[header];
      } else {
        let value: unknown = item;
        const headerParts = header.split('.');
        for (const key of headerParts) {
          if (Array.isArray(value)) {
            value = JSON.stringify(
              value.map(item => {
                const v = item as Record<string, unknown>;
                return `${(v[key] ? v[key] : '') as string}`;
              })
            );
          } else if (value !== null && typeof value === 'object') {
            value = value[key as keyof typeof value];
          }
        }
        const answer = stringifyResponse(value);
        rows[header] = answer;
      }
    });
    return rows;
  });
  return result;
}

const MAX_DEPTH = 3;

export type SourceField = {
  [key: string]: SourceField | null;
};

const retrieveSourceFieldsFromSchema = (
  type: GraphQLOutputType,
  options?: CoerceFieldsIntoOperationOptions,
  path = '',
  depth = 0
): SourceField => {
  if (depth >= (options?.maxNested ?? MAX_DEPTH)) {
    return {};
  }
  const result: SourceField = {};
  const namedType = getNamedType(type);

  if (isObjectType(namedType)) {
    const fields = namedType.getFields();
    for (const fieldName in fields) {
      const field = fields[fieldName];
      const currPath = path ? `${path}.${fieldName}` : fieldName;
      const fieldNamedType = getNamedType(field.type);
      const isLeaf = isScalarType(fieldNamedType);
      const isObject = isObjectType(fieldNamedType);

      const isIncluded =
        options?.include == null ||
        options.include.some(includePath =>
          isLeaf
            ? currPath.startsWith(`${includePath}.`) || currPath === includePath
            : includePath.startsWith(currPath)
        );

      const isExcluded = options?.exclude ? options.exclude.includes(currPath) : false;
      if (isIncluded && !isExcluded) {
        if (isObject) {
          const nestedObject = retrieveSourceFieldsFromSchema(
            field.type,
            options,
            currPath,
            depth + 1
          );
          if (Object.keys(nestedObject).length > 0) {
            result[fieldName] = nestedObject;
          }
        } else {
          result[fieldName] = null;
        }
      }
    }
  }

  return result;
};

export const getSourceFields = (
  schema: GraphQLSchema,
  source: string,
  options?: CoerceFieldsIntoOperationOptions
): SourceField => {
  const rootType = findOperationRootTypeOrThrow(schema, 'query', source);
  const namedType = getNamedType(rootType);
  const fields = isObjectType(namedType) ? Object.entries(namedType.getFields()) : [];
  const items = fields.filter(([fieldName]) => {
    return fieldName === 'items';
  });
  if (items.length > 0) {
    const itemsType = items[0][1].type;
    return retrieveSourceFieldsFromSchema(itemsType, options);
  }
  return {};
};