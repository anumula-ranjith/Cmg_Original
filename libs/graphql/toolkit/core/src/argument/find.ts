import type { GraphQLArgument, GraphQLSchema } from 'graphql';
import { isObjectType } from 'graphql';

/**
 * Finds the GraphQL arguments from an ObjectType's field.
 * @param schema The schema to explore.
 * @param typeName The GraphQL Type name name.
 * @param field The field name of the GraphQL Typ.
 * @returns An array of GraphQL Arguments.
 */
export const findArgumentsByField = (
  schema: GraphQLSchema,
  typeName: string,
  field: string
): readonly GraphQLArgument[] => {
  const type = schema.getType(typeName);
  if (!isObjectType(type)) {
    return [];
  }

  const fields = type.getFields();
  const args = Object.values(fields).find(fieldType => fieldType.name === field)?.args;
  return args ?? [];
};
