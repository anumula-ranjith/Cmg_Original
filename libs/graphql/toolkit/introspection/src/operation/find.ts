import type { Maybe } from '@cmg/core-types';
import type {
  ASTKindToNode,
  GraphQLOutputType,
  GraphQLSchema,
  OperationDefinitionNode,
  OperationTypeNode,
  Visitor,
} from 'graphql';
import { TypeInfo, visit, visitWithTypeInfo } from 'graphql';

/**
 * Finds a Root Operation (from Query or Mutation) GraphQL Type.
 * @param schema The GraphQL Schema.
 * @param operationType The Operation Type.
 * @param field The field name.
 * @returns A GraphQL Type if the type found, undefined otherwise.
 */
export const findOperationRootType = (
  schema: GraphQLSchema,
  operationType: OperationTypeNode,
  field: string | OperationDefinitionNode
): Maybe<GraphQLOutputType> => {
  const typeInfo = new TypeInfo(schema);
  let outputType: Maybe<GraphQLOutputType> = undefined;
  const visitor: Visitor<ASTKindToNode> = {
    enter(node) {
      outputType = typeInfo.getType();
      typeInfo.enter(node);
    },
    leave(node) {
      typeInfo.leave(node);
    },
  };
  visit(
    typeof field === 'string'
      ? {
          kind: 'OperationDefinition',
          operation: operationType,
          selectionSet: {
            kind: 'SelectionSet',
            selections: [{ kind: 'Field', name: { kind: 'Name', value: field } }],
          },
        }
      : field,
    visitWithTypeInfo(typeInfo, visitor)
  );
  return outputType;
};

/**
 * Finds a Root Operation (from Query or Mutation) GraphQL Type. Throws if the type is not found.
 * @param schema The GraphQL Schema.
 * @param operationType The Operation Type.
 * @param field The field name.
 * @returns A GraphQL Type if the type found, throws otherwise.
 */
export const findOperationRootTypeOrThrow = (
  schema: GraphQLSchema,
  operationType: OperationTypeNode,
  field: string | OperationDefinitionNode
): GraphQLOutputType => {
  const type = findOperationRootType(schema, operationType, field);
  if (type == null) {
    throw new Error(
      `The operation type root ${
        typeof field === 'string' ? field : field.name?.value ?? '<unknown>'
      } was not found in the schema.`
    );
  }

  return type;
};
