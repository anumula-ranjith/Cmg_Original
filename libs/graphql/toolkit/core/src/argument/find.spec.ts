/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { IntrospectionQuery } from 'graphql';
import { buildClientSchema } from 'graphql';

import introspectionJson from './__mocks__/introspection.json';
import { findArgumentsByField } from './find';

const introspectSchema = introspectionJson.data as unknown as IntrospectionQuery;

describe('findArgumentsByField', () => {
  it('find arguments for a Query type', () => {
    const schema = buildClientSchema(introspectSchema);
    const typeName = schema.getQueryType()?.name;
    expect(typeName).not.toBeFalsy();
    const result = findArgumentsByField(schema, typeName!, 'offerings');

    expect(result).toHaveLength(4);
    expect(result).toMatchSnapshot();
  });

  it('finds arguments for a Mutation type', () => {
    const schema = buildClientSchema(introspectSchema);
    const typeName = schema.getMutationType()?.name;
    expect(typeName).not.toBeFalsy();
    const result = findArgumentsByField(schema, typeName!, 'addCompany');

    expect(result).toMatchSnapshot();
  });
});
