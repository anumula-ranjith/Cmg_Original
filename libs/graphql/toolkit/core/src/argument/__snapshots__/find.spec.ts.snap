// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findArgumentsByField find arguments for a Query type 1`] = `
Array [
  Object {
    "astNode": undefined,
    "defaultValue": undefined,
    "deprecationReason": undefined,
    "description": null,
    "extensions": undefined,
    "name": "skip",
    "type": "Int",
  },
  Object {
    "astNode": undefined,
    "defaultValue": undefined,
    "deprecationReason": undefined,
    "description": null,
    "extensions": undefined,
    "name": "take",
    "type": "Int",
  },
  Object {
    "astNode": undefined,
    "defaultValue": undefined,
    "deprecationReason": undefined,
    "description": null,
    "extensions": undefined,
    "name": "order",
    "type": "[OfferingSortInput!]",
  },
  Object {
    "astNode": undefined,
    "defaultValue": undefined,
    "deprecationReason": undefined,
    "description": null,
    "extensions": undefined,
    "name": "where",
    "type": "OfferingFilterInput",
  },
]
`;

exports[`findArgumentsByField finds arguments for a Mutation type 1`] = `
Array [
  Object {
    "astNode": undefined,
    "defaultValue": undefined,
    "deprecationReason": undefined,
    "description": null,
    "extensions": undefined,
    "name": "input",
    "type": "AddCompanyInput!",
  },
]
`;
